pulser/simulation/simulation.py

```diff
@@ -297,57 +337,222 @@ def get_hamiltonian(self, time):
-            meas_basis = self._seq._measurement
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
+
+        def _run_solver(as_subroutine=False,
+                        measurement_basis=None):
+            """Returns:
+                CleanResults: Object containing the time evolution results. Its
+                _states attribute contains QuTiP quantum states, not Counters.
+            """
+            if not as_subroutine:
+                # CLEAN SIMULATION:
+                # Build (clean) hamiltonian
+                self._build_hamiltonian()
+                measurement_basis = _assign_meas_basis()
+
+            result = qutip.mesolve(self._hamiltonian,
+                                   self._config['initial_state'],
+                                   self.eval_times,
+                                   c_ops=self._collapse_ops,
+                                   progress_bar=progress_bar,
+                                   options=qutip.Options(max_step=5,
+                                                         **options)
+                                   )
+            return CleanResults(result.states, self.dim, self._size,
+                                self.basis_name, measurement_basis)
+
+        if self._noise:
+            # NOISY SIMULATION:
+            # We run the system multiple times
+            time_indices = range(len(self.eval_times))
+            total_count = np.array([Counter() for _ in time_indices])
+            for _ in range(self._config['runs']):
+                # At each run, new random noise
+                if 'SPAM' in self._noise:
+                    self._prepare_spam_detune()
+                self._build_hamiltonian()
+                meas_basis = _assign_meas_basis()
+                # Get CleanResults instance from sequence with added noise:
+                res_with_noise = _run_solver(as_subroutine=True,
+                                             measurement_basis=meas_basis)
+                # Extract statistics at eval time (final time by default):
+                if 'SPAM' in self._noise:
+                    total_count += \
+                        [res_with_noise.sampling_with_detection_errors(
+                            self.spam_dict,
+                            t=t,
+                            meas_basis=meas_basis,
+                            N_samples=self._config['samples_per_run'])
+                         for t in time_indices]
+                else:
+                    total_count += \
+                        [res_with_noise.sample_state(
+                            t=t,
+                            meas_basis=meas_basis,
+                            N_samples=self._config['samples_per_run'])
+                         for t in time_indices]
+            prob = [Counter({k: v / (self._config['runs']
```

Comment:
This could also be more explicitly named. What about instead of `total_count` above we call it `single_run_count` and instead of `prob` we call this `total_run_count`?